### PR TITLE
Update docs & notes

### PR DESCRIPTION
**Pull request by hanlint**
This PR:
* adds a `Notes` section to the docs
* adds a note about a cool feature -> automatic gradient accumulation

### CODE REVIEW

**Review by ravi-mosaicml — APPROVED** [review]
Overall LGTM -- Easy to read and follow!

Would it be possible to link to this doucment in https://mosaicml-composer--1164.com.readthedocs.build/en/1164/trainer/using_the_trainer.html?highlight=gradient#gradient-accumulation, and in the trainer docstrings for the `grad_accum` argument?

Looking forward to seeing the notebook :)

**Comment by hanlint** [comment]
Update: this PR requires a re-review. It now also:
* adds a note on autoresumption
* shortens some of the notebook names
* moves some of the topics (e.g. precision) from trainer -> notes. Not sure the right demarcation here.

File: docs/source/index.rst [hunk 1]
```diff
@@ -85,10 +85,17 @@ Composer is part of the broader Machine Learning community, and we welcome any c
...
-   trainer/distributed_training.rst
-   trainer/numerics.md
    trainer/performance.rst
 
+.. toctree::
+   :hidden:
+   :maxdepth: 1
+   :caption: Notes
+
+   trainer/distributed_training.rst
```
Comment (ravi-mosaicml) [inline docs/source/index.rst:95]: Would it be possible to move these files to the "notes" folder to match the hierarchy in the index file?

File: docs/source/index.rst [hunk 2]
```diff
@@ -49,7 +49,7 @@ Composer is part of the broader Machine Learning community, and we welcome any c
 .. toctree::
    :hidden:
    :maxdepth: 1
-   :caption: Examples
+   :caption: Notebooks
```
Comment (ravi-mosaicml) [inline docs/source/index.rst:52]: Even though they are notebooks, would prefer to keep them called "examples" or "tutorials"  -- I think that is the more standard terminology (pytorch calls them tutorials; ptl has tutorials and examples)

File: docs/source/notes/resumption.rst [hunk 3]
```diff
@@ -0,0 +1,134 @@
+|:play_or_pause_button:| Auto Resumption
+========================================
+
+Resuming from checkpoints is commonly used to recover from hardware failures (e.g. spot instances in the cloud being terminiated), loss spikes for large language models, or other unforseen errors. Our trainer supports resuming from checkpoints with the ``load_path`` argument (see :doc:`/trainer/checkpointing` for more details).
+
+.. testsetup::
```
Comment (ravi-mosaicml) [inline docs/source/notes/resumption.rst:6]: should this be the `.. testcode::`?

File: docs/source/notes/resumption.rst [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+
+    assert os.path.exists("./path/to/checkpoints/ep1.pt")
+
+    if not os.path.exists("./path/to/checkpoints/ep25.pt"):
+        shutil.copy2("./path/to/checkpoints/ep1.pt", "./path/to/checkpoints/ep25.pt")
+
+    assert os.path.exists("./path/to/checkpoints/ep25.pt")
+
+
+.. testcode::
```
Comment (ravi-mosaicml) [inline docs/source/notes/resumption.rst:32]: I presume this is from debugging?

File: docs/source/notes/resumption.rst [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+    trainer = Trainer(
+        ...,
+        autoresume=True,
+        save_folder='./path/to/folder',
+        run_name='my_cool_run',
+    )
+
+With autoresume, users can re-submit the _same_ code to the training run, and the trainer will handle finding and resuming from the latest checkpoints. This works well with systems like Kubernetes that automatically resubmit the same job when there is a node failure (due to spot instances as well). For ``autoresume=True`` to work, we require that both a ``save_folder`` and a ``run_name`` be provided. These are used to search for existing checkpoints.
+
+For an example code, see ``examples/checkpoint_autoresume.py``.
```
Comment (ravi-mosaicml) [inline docs/source/notes/resumption.rst:51]: (This can be done in a future PR), but I would like to convert this example script into a notebook. For now, can this be a link to the file on the dev branch?

File: docs/source/notes/resumption.rst [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        save_filename="ep{epoch}.pt",
+        save_folder="./path/to/checkpoints",
+        save_overwrite=True,
+        save_interval="1ep",  # Save checkpoints every epoch
+    )
+    trainer.fit()
+
+    assert os.path.exists("./path/to/checkpoints/ep1.pt")
+
+    if not os.path.exists("./path/to/checkpoints/ep25.pt"):
```
Comment (ravi-mosaicml) [inline docs/source/notes/resumption.rst:26]: Curious why the copy here?

File: docs/source/notes/resumption.rst [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+    )
+
+With autoresume, users can re-submit the _same_ code to the training run, and the trainer will handle finding and resuming from the latest checkpoints. This works well with systems like Kubernetes that automatically resubmit the same job when there is a node failure (due to spot instances as well). For ``autoresume=True`` to work, we require that both a ``save_folder`` and a ``run_name`` be provided. These are used to search for existing checkpoints.
+
+For an example code, see ``examples/checkpoint_autoresume.py``.
+
+Implementation
+--------------
+
+During training, the trainer always symlinks the latest checkpoint to a format (default is ``latest-rank{rank}`` for local files and ``{run_name}/checkpoints/latest-rank{rank}`` for object stores). When ``autoresume=True``, the Trainer searches for checkpoints of that format in the following order:
```
Comment (ravi-mosaicml) [inline docs/source/notes/resumption.rst:56]: This "Implementation" section is 100% accurate but very dense. Perhaps would it be possible to include an example for each of these scenarios?

File: docs/source/notes/resumption.rst [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+
+A typical use case is saving checkpoints to object store (e.g. S3) when there is no local file storage shared across runs. For example, a setup such as this:
+
+
+.. testcode::
+
+    from composer.loggers import ObjectStoreLogger
+
+    object_store_logger = ObjectStoreLogger(
+        object_store_cls=LibcloudObjectStore,
```
Comment (ravi-mosaicml) [inline docs/source/notes/resumption.rst:73]: For S3, would it be possible to use the S3ObjectStore class? That will automatically load creds from the environment.

File: docs/source/notes/resumption.rst [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        },
+    )
+
+.. testcode::
+
+    trainer = Trainer(
+        ...,
+        autoresume=True,
+        load_path='pretrained_weights/model.pt',
+        save_path='checkpoints',
```
Comment (ravi-mosaicml) [inline docs/source/notes/resumption.rst:126]: Would it be helpful to include the "load_object_store" argument to show that the pretrained weights could be stored in a bucket too?

**Review by ravi-mosaicml — APPROVED** [review]
Overall LGTM! Thanks for writing this up.

Feel free to take or leave any of the comments; nothing major.

File: docs/source/index.rst (hunk 2 above)
Comment (hanlint) [reply docs/source/index.rst:52]: Changed to Tutorials

File: docs/source/notes/resumption.rst (hunk 3 above)
Comment (hanlint) [reply docs/source/notes/resumption.rst:6]: fixed

File: docs/source/notes/resumption.rst (hunk 4 above)
Comment (hanlint) [reply docs/source/notes/resumption.rst:32]: Ah yes, fixed

File: docs/source/notes/resumption.rst (hunk 5 above)
Comment (hanlint) [reply docs/source/notes/resumption.rst:51]: Done

File: docs/source/notes/resumption.rst (hunk 6 above)
Comment (hanlint) [reply docs/source/notes/resumption.rst:26]: The reason is that for fast tests, we don't actually train for 25 epochs, so want to pretend the ep1 checkpoint is the ep25 checkpoint.

This code is actually copied from `resumption.rst`, and based on code in this PR: https://github.com/mosaicml/composer/commit/d025644e5786a84ac7edebbc6284813df1503c98.

File: docs/source/notes/resumption.rst (hunk 7 above)
Comment (hanlint) [reply docs/source/notes/resumption.rst:56]: The examples following illustrate 2 and 3. I added some text to clarify.

File: docs/source/notes/resumption.rst (hunk 8 above)
Comment (hanlint) [reply docs/source/notes/resumption.rst:73]: Done

File: docs/source/notes/resumption.rst (hunk 9 above)
Comment (hanlint) [reply docs/source/notes/resumption.rst:126]: I added a note about this, so as not to clutter up this signature too much.
